### ISSUE TITLE
ARM template fail on resourceId() & Subscription's Topic dependency

### ISSUE THREAD

**Issue by ApiO** [Azure/azure-quickstart-templates#3556]
Hi,

I 've the following template describing a service bus, a topic and its subscription as follow: 

```
{
    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
		"service_bus_name": {"type":"string", "defaultValue":"myservicebus"},
		"topic_name":       {"type":"string", "defaultValue":"mytopic"},
		"subscription_name":{"type":"string", "defaultValue":"mysubscription"}
	},
    "variables": {
        "api_version":       "2015-08-01",
        "location":          "[resourceGroup().location]",	
        "service_bus_name":  "[parameters('service_bus_name')]",
        "topic_name":        "[concat(variables('service_bus_name'), '/', parameters('topic_name'))]",
        "subscription_name": "[concat(variables('topic_name'), '/', parameters('subscription_name'))]"
	},
    "resources": [
        {
            "type": "Microsoft.ServiceBus/namespaces",
            "name": "[variables('service_bus_name')]",
            "apiVersion": "[variables('api_version')]",
            "location": "[variables('location')]",
            "sku": {
                "name": "Standard",
                "tier": "Standard"
            },
            "kind": "Messaging",
            "tags": {},
            "properties": {
                "provisioningState": "Succeeded",
                "status": "Active",
                "serviceBusEndpoint": "[concat('https://', variables('service_bus_name'),'.servicebus.windows.net:443/')]",
                "enabled": true,
                "createdAt": null,
                "updatedAt": null
            },
            "dependsOn": []
        },
        {
            "type": "Microsoft.ServiceBus/namespaces/topics",
            "name": "[variables('topic_name')]",
            "apiVersion": "[variables('api_version')]",
            "location": "[variables('location')]",
            "properties": {
                "defaultMessageTimeToLive": "14.00:00:00",
                "maxSizeInMegabytes": 5120,
                "requiresDuplicateDetection": false,
                "enableBatchedOperations": true,
                "sizeInBytes": 0,
                "filteringMessagesBeforePublishing": false,
                "isAnonymousAccessible": false,
                "status": "Active",
                "createdAt": null,
                "updatedAt": null,
                "accessedAt": null,
                "supportOrdering": false,
                "countDetails": {
                    "activeMessageCount": 0,
                    "deadLetterMessageCount": 0,
                    "scheduledMessageCount": 0,
                    "transferMessageCount": 0,
                    "transferDeadLetterMessageCount": 0
                },
                "subscriptionCount": 1,
                "autoDeleteOnIdle": "10675199.02:48:05.4775807",
                "enablePartitioning": true,
                "isExpress": false,
                "enableSubscriptionPartitioning": false,
                "enableExpress": false
            },
            "dependsOn": [
                "[resourceId('Microsoft.ServiceBus/namespaces', variables('service_bus_name'))]"
            ]
        },
        {
            "type": "Microsoft.ServiceBus/namespaces/topics/subscriptions",
            "name": "[variables('subscription_name')]",
            "apiVersion": "[variables('api_version')]",
            "location": "[variables('location')]",
            "scale": null,
            "properties": {
                "lockDuration": "00:00:30",
                "requiresSession": false,
                "defaultMessageTimeToLive": "14.00:00:00",
                "deadLetteringOnMessageExpiration": false,
                "deadLetteringOnFilterEvaluationExceptions": false,
                "messageCount": 0,
                "maxDeliveryCount": 10,
                "enableBatchedOperations": false,
                "status": "Active",
                "createdAt": null,
                "updatedAt": null,
                "accessedAt": null,
                "countDetails": { 
                    "activeMessageCount": 0,
                    "deadLetterMessageCount": 0,
                    "scheduledMessageCount": 0,
                    "transferMessageCount": 0,
                    "transferDeadLetterMessageCount": 0
                },
                "autoDeleteOnIdle": "10675199.02:48:05.4775807",
                "entityAvailabilityStatus": "Available"
            },
            "dependsOn": [
                "[resourceId('Microsoft.ServiceBus/namespaces', variables('service_bus_name'))]",
                "[resourceId('Microsoft.ServiceBus/namespaces/topics', variables('topic_name'))]"
            ]
        }
    ]
}
```
Code structure came from an automation script generated by the Portal, with some tweaking on parameters/variables.

When I deploy I've the following error:

```
{
	"telemetryId": "e2e1c79c-0e33-4ba1-a80b-0203ba2603a1",
	"galleryItemId": "Microsoft.Template",
	"createBlade": "DeployToAzure",
	"code": "InvalidTemplate",
	"message": "Deployment template validation failed: 'The template resource 'myservicebus/mytopic/mysubscription' at line '1' and column '2164' is not valid: Unable to evaluate template language function 'resourceId': function requires exactly one multi-segmented argument which must be resource type including resource provider namespace. Current function arguments 'Microsoft.ServiceBus/namespaces/topics,myservicebus/mytopic'. Please see https://aka.ms/arm-template-expressions/#resourceid for usage details.. Please see https://aka.ms/arm-template-expressions for usage details.'."
}
```
When I deploy my orginal script, without tweaking, I have the exact same error.

I checked the [resourceId function spec](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-template-functions-resource#resourceid), and everything seems to be correct.

I also tried  to set explicitly the not required params, as:
```
[resourceId(subscription().subscriptionId, resourceGroup().name, 'Microsoft.ServiceBus/namespaces/topics', variables('topic_name'))]
```
... same error.

I took a look to the [101-servicebus-topic-subscription template](https://github.com/Azure/azure-quickstart-templates/tree/master/101-servicebus-topic-subscription), but it dosen't work either.

I removed the subscription resource to validate the others resources script and everything works.

I hope to miss something huge, I'm near to make a clean CI/CD solution, it's very frustrating ... but template seems great !! Damn !

**Comment by krnese** [comment]
It looks like you have mixed-up your template with some wrong variables. 

See if this works:

{
    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "service_bus_name": {
            "type": "string",
            "defaultValue": "myservicebus"
        },
        "topic_name": {
            "type": "string",
            "defaultValue": "mytopic"
        },
        "subscription_name": {
            "type": "string",
            "defaultValue": "mysubscription"
        }
    },
    "variables": {
        "api_version": "2015-08-01",
        "location": "[resourceGroup().location]",
        "service_bus_name": "[parameters('service_bus_name')]",
        "topic_name": "[concat(variables('service_bus_name'), '/', parameters('topic_name'))]",
        "subscription_name": "[concat(variables('topic_name'), '/', parameters('subscription_name'))]"
    },
    "resources": [
        {
            "type": "Microsoft.ServiceBus/namespaces",
            "name": "[variables('service_bus_name')]",
            "apiVersion": "[variables('api_version')]",
            "location": "[variables('location')]",
            "sku": {
                "name": "Standard",
                "tier": "Standard"
            },
            "kind": "Messaging",
            "tags": {},
            "properties": {
                "provisioningState": "Succeeded",
                "status": "Active",
                "serviceBusEndpoint": "[concat('https://', variables('service_bus_name'),'.servicebus.windows.net:443/')]",
                "enabled": true,
                "createdAt": null,
                "updatedAt": null
            },
            "dependsOn": []
        },
        {
            "type": "Microsoft.ServiceBus/namespaces/topics",
            "name": "[variables('topic_name')]",
            "apiVersion": "[variables('api_version')]",
            "location": "[variables('location')]",
            "properties": {
                "defaultMessageTimeToLive": "14.00:00:00",
                "maxSizeInMegabytes": 5120,
                "requiresDuplicateDetection": false,
                "enableBatchedOperations": true,
                "sizeInBytes": 0,
                "filteringMessagesBeforePublishing": false,
                "isAnonymousAccessible": false,
                "status": "Active",
                "createdAt": null,
                "updatedAt": null,
                "accessedAt": null,
                "supportOrdering": false,
                "countDetails": {
                    "activeMessageCount": 0,
                    "deadLetterMessageCount": 0,
                    "scheduledMessageCount": 0,
                    "transferMessageCount": 0,
                    "transferDeadLetterMessageCount": 0
                },
                "subscriptionCount": 1,
                "autoDeleteOnIdle": "10675199.02:48:05.4775807",
                "enablePartitioning": true,
                "isExpress": false,
                "enableSubscriptionPartitioning": false,
                "enableExpress": false
            },
            "dependsOn": [
                "[resourceId('Microsoft.ServiceBus/namespaces/', variables('service_bus_name'))]"
            ]
        },
        
    ]
}

**Comment by ApiO** [comment]
After further experiments I managed to make the template working by changing the resourceId() function usage:

```
"[resourceId('Microsoft.ServiceBus/namespaces/topics', parameter('service_bus_name'), parameter('topic_name'))]"
```

I also changed the api version to 2017-04-01.

Why does it work ? The function documentation is not completly clear around the description:

> Parameter: resourceName2	
> Required: No	
> Type: string	
> Description: Next resource name segment if resource is nested.

I think my template now work because the "nested thing" is setup, but why I have to manage that when I'm currently dealing with dependency declaration ? Specify resourceType & resourceName should be enough, no ? 


@krnese thanks for your reply. I haven't tried your script because, first, it seems good :) but last, it does not contains the subscription resource part. I already tried that and it work.
